File: src/main/java/co/uk/isxander/evergreenhud/elements/impl/ElementBlockAbove.java
/*
 * Copyright (C) Evergreen [2020 - 2021]
 * This program comes with ABSOLUTELY NO WARRANTY
 * This is free software, and you are welcome to redistribute it
 * under the certain conditions that can be found here
 * https://www.gnu.org/licenses/lgpl-3.0.en.html
 *
 * If you have any questions or concerns, please create
 * an issue on the github page that can be found here
 * https://github.com/Evergreen-Client/EvergreenHUD
 *
 * If you have a private concern, please contact
 * isXander @ business.isxander@gmail.com
 */

package co.uk.isxander.evergreenhud.elements.impl;

import co.uk.isxander.evergreenhud.elements.type.SimpleTextElement;
import co.uk.isxander.evergreenhud.settings.impl.BooleanSetting;
import co.uk.isxander.evergreenhud.settings.impl.IntegerSetting;
import co.uk.isxander.evergreenhud.elements.ElementData;
import com.google.common.collect.Sets;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.init.Blocks;
import net.minecraft.util.BlockPos;
import net.minecraftforge.event.entity.living.LivingEvent;

import java.util.HashSet;
import java.util.Set;

public class ElementBlockAbove extends SimpleTextElement {

    private static final Set<Block> EXEMPT_BLOCKS = Sets.newHashSet(Blocks.air, Blocks.water, Blocks.lava, Blocks.ladder, Blocks.vine, Blocks.wall_sign, Blocks.wall_sign, Blocks.standing_sign, Blocks.standing_banner);
    private int blockDistance = 0;

    public BooleanSetting notify;
    public IntegerSetting notifyHeight;
    public IntegerSetting checkAmount;

    @Override
    public void initialise() {
        addSettings(notify = new BooleanSetting("Notify", "Sound", "Make a noise when the block gets too close.", false));
        addSettings(notifyHeight = new IntegerSetting("Notify Height", "Functionality", "How close the block needs to be before notifying.", 3, 1, 10, " blocks"));
        addSettings(checkAmount = new IntegerSetting("Check Amount", "Functionality", "How many blocks the element measures before stopping. (Turning this to a high value may cause lag.)", 10, 1, 30, " blocks"));
    }

    @Override
    public ElementData metadata() {
        return new ElementData("Block Above", "Tells you if there is a block above your head. Useful for games like bedwars.", "Advanced");
    }

    @Override
    protected String getValue() {
        return Integer.toString(blockDistance);
    }

    @Override
    public void onLivingUpdate(LivingEvent.LivingUpdateEvent event) {
        if (mc.theWorld == null) return;
        if (event.entity != mc.thePlayer) return;

        boolean above = false;
        for (int i = 1; i < checkAmount.get() + 1; i++) {
            BlockPos blockPos = new BlockPos(mc.thePlayer.posX, mc.thePlayer.posY + 1 + i, mc.thePlayer.posZ);
            if (blockPos.getY() > 255) break;

            IBlockState state;
            try {
                state = mc.theWorld.getBlockState(blockPos);
            } catch (NullPointerException e) {
                continue;
            }

            if (state == null) continue;
            Block b = state.getBlock();
            if (!EXEMPT_BLOCKS.contains(b)) {
                if (i <= notifyHeight.get() && (blockDistance > notifyHeight.get() || blockDistance == 0)) {
                    if (notify.get())
                        mc.thePlayer.playSound("random.orb", 0.1f, 0.5f);
                }
                blockDistance = i;
                above = true;
                break;
            }
        }
        if (!above)
            blockDistance = 0;
    }

    @Override
    public String getDefaultDisplayTitle() {
        return "Above";
    }
}

File: src/main/kotlin/dev/isxander/evergreenhud/elements/impl/ElementBlockAbove.kt
/*
 * EvergreenHUD - A mod to improve your heads-up-display.
 * Copyright (c) isXander [2019 - 2022].
 *
 * This work is licensed under the GPL-3 License.
 * To view a copy of this license, visit https://www.gnu.org/licenses/gpl-3.0.en.html
 */

package dev.isxander.evergreenhud.elements.impl

import dev.isxander.evergreenhud.elements.type.SimpleTextElement
import dev.isxander.evergreenhud.utils.elementmeta.ElementMeta
import dev.isxander.evergreenhud.utils.mc
import dev.isxander.settxi.impl.boolean
import dev.isxander.settxi.impl.int
import net.minecraft.block.*
import net.minecraft.util.BlockPos


@ElementMeta(id = "evergreenhud:block_above", name = "Block Above", description = "Show how many blocks until you hit your head on something above you.", category = "World")
class ElementBlockAbove : SimpleTextElement("Above") {
    var notify by boolean(true) {
        name = "Notify"
        description = "Make a sound when you get close to hitting your head."
        category = "Block Above"
    }
    var notifyHeight by int(3) {
        name = "Notify Height"
        description = "How close you need to be before you notify."
        category = "Block Above"
        range = 1..10

        depends { notify }
    }
    var checkHeight by int(10) {
        name = "Check Height"
        description = "How many blocks until element stops checking blocks."
        category = "Block Above"
        range = 1..30
    }

    private var notified = false

    override fun calculateValue(): String {
        if (mc.theWorld == null || mc.thePlayer == null) return "0"

        var above = 0
        for (i in 1..checkHeight) {
            val pos = BlockPos(mc.thePlayer.posX, mc.thePlayer.posY + 1 + i, mc.thePlayer.posZ)
            if (pos.y > 255) break

            val state = mc.theWorld!!.getBlockState(pos) ?: continue
            if (state.block is BlockAir
                || state is BlockStaticLiquid
                || state.block is BlockSign
                || state is BlockVine
                || state.block is BlockBanner)
                continue

            above = i

            if (above <= notifyHeight && notify) {
                if (!notified) {
                    mc.thePlayer.playSound("random.orb", 0.1f, 0.5f)
                    notified = true
                }
            } else {
                notified = false
            }

            break
        }

        return above.toString()
    }
}
